sdb/coerce.py

```diff
@@ -65,7 +65,8 @@ def coerce(self, obj: drgn.Object) -> drgn.Object:
 
         # "void *" can be coerced to any pointer type
         if (obj.type_.kind is drgn.TypeKind.POINTER and
-                obj.type_.primitive is drgn.PrimitiveType.C_VOID):
+                (obj.type_.primitive is None or
+                 obj.type_.primitive is drgn.PrimitiveType.C_VOID)):
```

Comment:
I was a bit surprised for the same reasons that @prakashsurya pointed out above and then I took closer look yet I'm still curious.

So here are all the primitive types according to `drgn` (which I assume come straight from the DWARF specification):
```
class drgn.PrimitiveType¶
PrimitiveType is a enum.Enum of the primitive types known to drgn.

C_VOID
C_CHAR
C_SIGNED_CHAR
C_UNSIGNED_CHAR
C_SHORT
C_UNSIGNED_SHORT
C_INT
C_UNSIGNED_INT
C_LONG
C_UNSIGNED_LONG
C_LONG_LONG
C_UNSIGNED_LONG_LONG
C_BOOL
C_FLOAT
C_DOUBLE
C_LONG_DOUBLE
C_SIZE_T
C_PTRDIFF_T
```
According to the above there is no primitive type for a pointer implying that pointers are not primitive types. Double checking this in the `drgn` REPL we can verify it:
```
>>> a = drgn.Object(prog, type="int *", value=0xff)
>>> a.type_.primitive == None
True
``` 

So looking at the code snippet that Matt changed I think that the original if-statement is always false and that's a bug from our crash-python to drgn transition. I think what we meant to do during our porting effort was this:
```
# "void *" can be coerced to any pointer type
if (obj.type_.kind is drgn.TypeKind.POINTER and
    obj.type_.type.primitive is drgn.PrimitiveType.C_VOID):   # attention on type_.type - reads as "our type is a pointer and the type of that pointer is void"
...
```

So if we wanted to coerce only `void *` pointers we should do the above. That said Matt raises an interesting point with his code change - shall we allow any kind of pointer to be coerced to what we want? I can see both pros and cons with that but for the case that we decide to go for it we should update the comment above that statement .

Reply:
awesome, thanks for figuring that out!

> shall we allow any kind of pointer to be coerced to what we want

IMO, no: use `cast` for that.